rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,826 @@
+that guarding against shape mismatches should be made a producer responsibility.
+This has been a subject of many informal conversations as well as one of the
+discussion items
+[at the StableHLO dynamism breakout session](https://drive.google.com/drive/u/1/folders/1fGqq8Tcebhcwq1KJqAZPDgXksRdqRvM2)
+during the OpenXLA Dev Summit in April 2023.
+
+**Proposal:** This RFC proposes to ratify this convention as a StableHLO design
+principle, so that the folklore consensus becomes codified in the StableHLO
+specification.
+
+**Discussion:** A) The rationale for this proposal is that lower-level
+abstractions (e.g. HLO or Linalg) typically don't want to concern themselves
+with shape checks - they are focused on generating high-performance code, so
+they want shape checks to be handled at a higher level, i.e. at StableHLO
+or above.
+
+Furthermore, different producers have different requirements and different
+preferences for expressing shape checks (e.g. JAX's type system enables it to
+need fewer checks than TensorFlow's type system), so a specific way of
+performing shape checks don't look like something that should be standardized
+at the StableHLO level either.
+
+B\) From the implementation standpoint, this proposal would involve updating
+[the "Errors" section](../docs/spec.md#errors) of the specification, and
+changing the StableHLO dialect to implement the `ConditionallySpeculatable`
+interface, prohibiting speculation for StableHLO ops that involve dynamism
+([documentation](https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/)).
+
+<!-- markdownlint-disable line-length -->
+## <a id="p3" /> (P3) Ratify the existing convention for relaxed constraints already implemented in the StableHLO dialect
+<!-- markdownlint-enable line-length -->
+
+**Baseline:** At the moment, the StableHLO dialect supports
+[relaxed constraints](https://github.com/openxla/stablehlo/blob/8993ad54839add6648b88801f1d223b7f9bc2e58/stablehlo/dialect/Base.cpp#L102-L120)
+that were inherited from the MHLO dialect. For example, the code below is valid
+in the StableHLO dialect:
+
+```mlir
+func.func @main(%arg0: tensor<?xf32>, %arg1: tensor<1xf32>) {
+  %0 = stablehlo.add %arg0, %arg0 : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32>
+  %1 = stablehlo.add %arg0, %arg0 : (tensor<?xf32>, tensor<?xf32>) -> tensor<1xf32>
+  %2 = stablehlo.add %arg0, %arg1 : (tensor<?xf32>, tensor<1xf32>) -> tensor<?xf32>
+  %3 = stablehlo.add %arg0, %arg1 : (tensor<?xf32>, tensor<1xf32>) -> tensor<1xf32>
+  %4 = stablehlo.add %arg1, %arg0 : (tensor<1xf32>, tensor<?xf32>) -> tensor<?xf32>
+  %5 = stablehlo.add %arg1, %arg0 : (tensor<1xf32>, tensor<?xf32>) -> tensor<1xf32>
+  %6 = stablehlo.add %arg1, %arg1 : (tensor<1xf32>, tensor<1xf32>) -> tensor<?xf32>
+  %7 = stablehlo.add %arg1, %arg1 : (tensor<1xf32>, tensor<1xf32>) -> tensor<1xf32>
+  func.return
+}
+```
+
+More formally, these relaxed constraints generalize the constraints that are
```

Comment:
This means that for dynamic cases we should always check for shape mismatches in kernels at runtime?

Reply:
Not really. P2 proposes that shape mismatches constitute undefined behavior at the StableHLO level, i.e. that StableHLO consumers are not obligated to perform shape checks in their kernel implementations (so these shape checks must happen elsewhere).

On a related note, per O3, defining a standard mechanism for shape checks is out of scope for this RFC. TF/XLA uses `shape.assuming` for these purposes, JAX is using `stablehlo.custom_call @shape_assertion` ([example](https://github.com/tensorflow/tensorflow/blob/40f01670c823de04fe8302fa9bc6987a376bbf4f/tensorflow/compiler/tests/xla_call_module_test.py#L464)), and we haven't discussed this from the PyTorch perspective yet.

So, overall, the shape checking story for StableHLO is in a bit of a chaos right now, and it will be desirable to figure it out in a follow-up RFC.

Reply:
Got it. So far we don't see such shape check ops in the generated stablehlo module, probably because right now JAX only support shape polymorphism? If we are going to have these shape checking ops in the stablehlo module, we need to handle them in our converter.

Reply:
@gnecula Can you tell more about about the current status and the future plans for shape checks in JAX lowerings?

Reply:
In the first phase, JAX will use dynamic shapes only when the intermediate shapes depend only on the **shapes** of the inputs, not on the value of the inputs (we call this shape polymorphism). JAX will try to do all the shape checking symbolically at lowering time, but when it is not possible it will insert `custom call shape_assertions`, which are expected to be discharged and removed once the input shapes are known.

In a latter phase JAX plans to support more dynamism, but we have not yet decided whether we'll need runtime shape assertions. Probably we will start by rejecting programs that we cannot verify statically to be well-shaped.